Engineering/Frontend/Mitigate blocking the main thread.md

```diff
@@ -0,0 +1,55 @@
+---
+tags: javascript, frontend performance, engineering/frontend
+author: Phat G Ha
+date: 2022-31-08
+---
+
+We all know that for web applications, perceived performance is highly importance for our users.
+For data intensive SPAs with CPU-intensive tasks, the single-threaded nature of Javascript starts to hinder the application's perceived performance when you cannot fit those CPU-intensive tasks into a 16.67 ms/frame window (translates to 60fps). That doesn't even account for code execution time of frameworks (React, Vue...etc), which then leaves you with approximately only a 10 ms/frame window or less to complete all your tasks on the main thread to make user interaction feel smooth and snappy.
+
+Fortunately for us, JavaScript provides a couple of tools to solve these problems with some new features: Threading (through Web Worker) and Coroutines (through generators).
+
+### Threading with Web Worker
+
+If you don't know what a Web Worker is, check out [[Parallelism in JavaScript]] for a quick introduction.
+
+With Web Worker we can offload a bunch of processing to another thread and it will not impact the performance of the main thread. Sounds perfect, but it comes with a cost: `serialization`.
+
+Because of the heavy sandboxing of JS environment, using web workers only works well if we have small to medium inputs and outputs. Because all of the data transported between the main thread and worker thread is going to be serialized, which blocks the main thread while that happens (unless you are using binary formats like Shared Array Buffer or Typed Arrays, which comes with other caveats that is outside of the scope of this note).
+
+If `serialization` is not a problem for your app then Web Worker will probably do a great job to enhance your app's performance and UX.
+
+For real time apps that require continuously recalculation of data when new data comes, you can implement a game-loop-like updater function inside the web worker code to periodically emit data to main thread once per an interval time limit. That pattern works well because it allows you to control how often main thread should receive new data and UI should rerender.
+
+### Coroutines with generators
```

Comment:
Looking forward to a demo 😎 (if you plan to)

Reply:
I can copy the worker code from nghenhan and make a codesandbox lol, will add tmr

Reply:
If you can do a demo for it this Friday, it'd be great ⭐